src/runtime/metal/metal_hardware_manager.cpp

```diff
@@ -0,0 +1,641 @@
+/*
+ * This file is part of hipSYCL, a SYCL implementation based on CUDA/HIP
+ *
+ * Copyright (c) 2023 Aksel Alpay
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <cassert>
+#include <cstdint>
+#include <string>
+#include <limits>
+
+// TODO: Make a special file to declare these implementations.
+// You must define these *before* including Metal.hpp.
+#define NS_PRIVATE_IMPLEMENTATION
+#define MTL_PRIVATE_IMPLEMENTATION
+#define CA_PRIVATE_IMPLEMENTATION
+
+#include <Metal/Metal.hpp>
+#include <IOKit/IOKitLib.h>
+#include <CoreFoundation/CoreFoundation.h>
+#include <sys/sysctl.h>
```

Comment:
> glue is for code that connects two or more of the hipSYCL components (runtime, compiler, headers). Where would metal-cpp be needed? If it's just needed for the runtime, you could also put it inside the metal directory of the runtime.

metal-cpp is just needed for compile-time. It's a wrapper around the Objective-C Metal binaries, which ship with macOS.

Declaring these `#defines` in some random source file doesn't make much sense. We could make `metal_cpp_implementation.cpp` in this `runtime/metal` directory. But how is this part of the runtime?

Reply:
I had assumed that this metal-cpp provides runtime functionality, such as hardware management, submitting tasks etc. In that case the only user would be the hipSYCL metal runtime backend, so it would make sense to put it there.

Os is that not the case? Is it something that is needed by kernels?

Reply:
> Or is that not the case? Is it something that is needed by kernels?

It might also export C++ symbols. For example, we might permit Metal interop with SYCL code. That requires exposing the `MTL::CommandQueue`, `MTL::Buffer` etc. objects through a C++ API. Some of our shipped headers must import metal-cpp.

Reply:
Do we need metal-cpp for that, or would it be sufficient to just forward declare the exported types, and require that the user include metal-cpp by themselves if they want to use such interop? This is what the other backends currently do for interop.

Reply:
That might be the best idea.